reps/2023-03-10-batch-inference.md

```diff
@@ -0,0 +1,150 @@
+
+@ericl
+
+## Design and Architecture
+
+## API Changes
+
+1. Introduce `map_preprocessor` API directly to Ray Datasets.
+
+```python
+def map_preprocessor(preprocessor: Preprocessor, batch_size: Optional[int], prefetch_batches: Optional[int]):
+  """Apply the transform of the provided preprocessor to batches of data.
+
+  Args:
+    preprocessor: A :class:`~ray.data.preprocessor.Preprocessor` used to transform batches of data. The provided preprocessor is
+      expected to already be fitted if it is fittable.
+    batch_size: Same as `map_batches`
+    prefetch_batches: Same as `map_batches`
+  """
+```
+
+2. Introduce a `map_predictor` API directly to Ray Datasets.
+
+```python
+def map_predictor(predictor: Union[Predictor, Callable[[], Predictor]], 
+                  batch_size: Optional[int], 
+                  feature_columns: Optional[List[str]] = None, 
+                  keep_columns: Optional[List[str]] = None, 
+                  prefetch_batches: Optional[int] = 0, 
+                  min_scoring_workers: int = 1,
+                  max_scoring_workers: Optional[int] = None,
+                  num_cpus_per_worker: Optional[int] = None,
+                  num_gpus_per_worker: Optional[int] = None, 
+                  **predict_kwargs):
+  """Predict on batches of data.
+
+  Args:
+    predictor: An instance of a Predictor to use for prediction. For Predictors that are not serializable, or are loading large models
+      or that are created from AIR Checkpoints, a function that creates a Predictor can be passed in instead. 
+    batch_size: Batch size to use for prediction.
+    feature_columns: List of columns in the preprocessed dataset to use for
+      prediction. Columns not specified will be dropped
+      from `data` before being passed to the predictor.
+      If None, use all columns in the preprocessed dataset.
+    keep_columns: List of columns in the preprocessed dataset to include
+        in the prediction result. This is useful for calculating final
+        accuracies/metrics on the result dataset. If None,
+        the columns in the output dataset will contain
+        just the prediction results.
+    min_scoring_workers: Minimum number of scoring actors.
+    max_scoring_workers: If set, specify the maximum number of scoring actors.
+    num_cpus_per_worker: Number of CPUs to allocate per scoring worker.
+      Set to 1 by default.
+    num_gpus_per_worker: Number of GPUs to allocate per scoring worker.
+      Set to 0 by default. If you want to use GPUs for inference, please
+      specify this parameter.
+    ray_remote_args: Additional resource requirements to request from ray.
+    predict_kwargs: Keyword arguments passed to the predictor's
+        ``predict()`` method.
+  """
+```
+
+3. Deprecate `Chain` Preprocessor, as chains can be represented via multiple `map_preprocessor` calls.
```

Comment:
For sake of keeping the scope narrow, I think we should not alter the preprocessor APIs in this REP.

Reply:
removed